src/main/java/org/wildfly/security/auth/provider/jdbc/JdbcSecurityRealm.java

```diff
@@ -111,13 +108,9 @@ public CredentialSupport getCredentialSupport(Class<?> credentialType) throws Re
         @Override
         public <C> C getCredential(Class<C> credentialType) throws RealmUnavailableException {
             for (QueryConfiguration configuration : JdbcSecurityRealm.this.queryConfiguration) {
-                for (ColumnMapper mapper : configuration.getColumnMappers()) {
-                    if (KeyMapper.class.isInstance(mapper)) {
-                        KeyMapper keyMapper = (KeyMapper) mapper;
-
-                        if (keyMapper.getKeyType().isAssignableFrom(credentialType)) {
-                            return executeAuthenticationQuery(configuration, resultSet -> (C) keyMapper.map(resultSet));
-                        }
+                for (KeyMapper keyMapper : configuration.getColumnMappers(KeyMapper.class)) {
+                    if (keyMapper.getKeyType().isAssignableFrom(credentialType)) {
+                        return executePrincipalQuery(configuration, resultSet -> (C) keyMapper.map(resultSet));
```

Comment:
I missed this earlier, but this looks like an unsafe cast.  You could make it safe by replacing ``(C) blah`` with ``credentialType.cast(blah)``.